sphinx/config.py

```diff
@@ -391,9 +370,76 @@ def convert_source_suffix(app, config):
+
+    See https://reproducible-builds.org/specs/source-date-epoch/
+    """
+    if getenv('SOURCE_DATE_EPOCH') is not None:
+        for k in ('copyright', 'epub_copyright'):
+            if k in config:
+                replace = r'\g<1>%s' % format_date('%Y')
+                config[k] = copyright_year_re.sub(replace, config[k])  # type: ignore
+
+
+def check_confval_types(app, config):
+    # type: (Sphinx, Config) -> None
+    """check all values for deviation from the default value's type, since
+    that can result in TypeErrors all over the place NB.
+    """
+    for confval in config:
+        settings = config.values[confval.name]
+        default = settings[0]
+        annotations = settings[2] if len(settings) == 3 else ()
+
+        if hasattr(default, '__call__'):
+            default = default(config)  # evaluate default value
+        if default is None and not annotations:
+            continue  # neither inferrable nor expliclitly annotated types
+
+        if annotations is Any:
+            # any type of value is accepted
+            pass
+        elif isinstance(annotations, ENUM):
+            if not annotations.match(confval.value):
+                msg = __("The config value `{name}` has to be a one of {candidates}, "
+                         "but `{current}` is given.")
+                logger.warning(msg.format(name=confval.name,
+                                          current=confval.value,
+                                          candidates=annotations.candidates))
+        else:
+            if type(confval.value) is type(default):
+                continue
+            if type(confval.value) in annotations:
+                continue
+
+            common_bases = (set(type(confval.value).__bases__ + (type(confval.value),)) &
+                            set(type(default).__bases__))
+            common_bases.discard(object)
+            if common_bases:
+                continue  # at least we share a non-trivial base class
+
+            if annotations:
+                msg = __("The config value `{name}' has type `{current.__name__}', "
+                         "expected to {permitted}.")
+                logger.warning(msg.format(name=confval.name,
+                                          current=type(confval.value),
+                                          permitted=str([c.__name__ for c in annotations])))
+            else:
+                msg = __("The config value `{name}' has type `{current.__name__}', "
+                         "defaults to `{default.__name__}'.")
+                logger.warning(msg.format(name=confval.name,
+                                          current=type(confval.value),
+                                          default=type(default)))
+
+
 def setup(app):
     # type: (Sphinx) -> Dict[unicode, Any]
     app.connect('config-inited', convert_source_suffix)
+    app.connect('config-inited', correct_copyright_year)
+    app.connect('config-inited', check_confval_types)
```

Comment:
This refactoring separates config-manipulators from `Config` class.
I believe they are not responsibility of the class.

Reply:
I'm not so sure about this one. Unlike `check_copyright_year` and `convert_source_suffix`, which are concerned with specific options, this one concerns the public API for adding options. It seems like this is less about manipulation and more about validation. Surely validation should remain part of the functionality of the class?